library/oeUI/src/OemrUI.php

```diff
@@ -0,0 +1,290 @@
+
+class OemrUI
+{
+    private $action;
+    private $action_href;
+    private $action_icon;
+    private $action_title;
+    private $arrAction;
+    private $arrow_direction;
+    private $arrexpandIcon;
+    private $arrOeUiSettings;
+    private $close;
+    private $container;
+    private $current_state;
+    private $display_help_icon;
+    private $expand_icon_class;
+    private $expand_title;
+    private $expandable;
+    private $expandable_icon;
+    private $heading;
+    private $help_file;
+    private $help_icon;
+    private $help_modal;
+    private $jquery_draggable;
+    private $modal_body;
+    private $print;
+    private $target;
+       
+    /**
+    * Used to create the html string that will display the formatted heading with selected icons - expandable, action and help and generate the html code for the help modal.
+    *
+    * @param array $arrOeUiSettings is an associative array that contains 9 elements, string 'heading_title', int|bool 'include_patient_name', int|bool 'expandable', array 'expandable_files',
+    * string 'action', string 'action_title', string 'action_href', int|bool 'show_help_icon' and string 'help_file_name'. The int|bool 'current_state' (expanded = 1, centered = 0)  value is obtained from function
+    * collectAndOrganizeExpandSetting(array("")), this function needs an indexed array as an argument (array 'expandable_files') that
+    * contains the file name of the current file as the first element, the name of any other file that needs to open in a similar state
+    * needs to be included in this array,all names must be unique and have a '_xpd' suffix.
+    * It will be used to generate up to 4 values - string $heading, string $expandable_icon, string $action_icon and string $help_icon that will form the html string used to output
+    * the formatted heading of the page.
+    * If a feature is not required set the corresponding element in the array to an empty string
+    */
+    public function __construct($arrOeUiSettings = array())
+    {
+        $this->heading = ($arrOeUiSettings['include_patient_name'])? $arrOeUiSettings['heading_title'] . " - " . getPatientNameFirstLast($_SESSION['pid']):$arrOeUiSettings['heading_title'];
+        $this->expandable = $arrOeUiSettings['expandable'];
+        $this->arrAction = array($arrOeUiSettings['action'], $arrOeUiSettings['action_title'], $arrOeUiSettings['action_href']);
+        $this->display_help_icon = $arrOeUiSettings['show_help_icon'];
+        $this->help_file = $arrOeUiSettings['help_file_name'];
+        if ($arrOeUiSettings['expandable'] && $arrOeUiSettings['expandable_files']) {
+            $this->current_state = collectAndOrganizeExpandSetting($arrOeUiSettings['expandable_files']);
+        }
+        echo "\r\n<script src='" . $GLOBALS['webroot'] . "/library/js/oeUI/universalTooltip.js?v=" . $v_js_includes. "'></script>\r\n";
```

Comment:
There are still 2 odd things going on that am trying to figure out :)
1. It's really odd to have constructor echo something like this(forcing folks to initiate the class in the right spot).
2. It's really odd to bring in assets (jquery-ui and the base theme) when initiate the class.(there will be clashing since developers will never know to look in the class; also bad things could happen if they want to bring in another jquery-ui theme rather than base)

Would this solution make sense:
1. Make it a requirement for scripts that use this class bring in via Header function the jquery-ui and a jquery ui theme (at least base). This will make sense to developers and will avoid clashing.
2. Then we can rig config.yml script to always bring in universalTooltip.js if jquery-ui is brought in

Reply:
I had structured it in this way to minimize adding extra code to get the class to work. As I had mentioned before `jquery-ui `was not being used in very many pages. We can certainly make it an autoloadable  feature in `config.yml`. I have not delved deeply into `jquery-ui ` but I did note that the `tooltip ` function of `bootstrap `and `jquery-ui ` differ in syntax. If `jquery-ui ` is included code using the` bootstrap tooltip ` syntax will fail. 

With regards initiating the class in the right spot I feel that placing the 

```
<?php
    $arrOeUiSettings = array(
        'heading_title' => xl('Access Control List Administration'),
        'include_patient_name' => false,
        'expandable' => false,
        'expandable_files' => array(),//all file names need suffix _xpd
        'action' => "link",//conceal, reveal, search, reset, link or back
        'action_title' => "",
        'action_href' => "../../gacl/admin/acl_admin.php",//only for actions - reset, link or back
        'show_help_icon' => true,
        'help_file_name' => "adminacl_help.php"
    );
    $oemr_ui = new OemrUI($arrOeUiSettings);
?>
```
code block as the last item before the `</head>` tag is intuitive and logical and thus the js that is being echoed will be in the `<head></head>` section of the page.

Furthermore I think it would be useful if we make it a convention to mark the end of the container div with 
`</div><!--end of container div-->`
and place 
`<?php $oemr_ui->helpFileModal();?>`
`<script> <?php require_once("$srcdir/js/oeUI/headerTitleAction.js"); ?></script>`
as the next two lines of code in order to 'standardize' the way the page source code will be rendered

Reply:
Hi,
You've convinced me :)
I like the end of container div idea.
Regarding the headerTitleAction.js, I noted some scripts use it and some don't; would it make sense to rename the helpFileModal() to something else like oeUIbottom which will do both the helpFileModal and /headerTitleAction.js on the same call (and can decide what to include)?

Reply:
Let me give it some thought. 
I kind of like `helpFileModal` as it conveys the purpose of the function. I suppose I could add` headerTitleAction.js.php` to the class but I feel leaving it as a js file and have the developer add these two as needed would minimize adding unnecessary code to every page.

Reply:
This is up to you.
My thoughts are that you are already telling the class if there is a helpfile and if it is expandable. So, might as well let the class decide what to add at with a call to the UiBottom() function, which will always go below the </div><!--end of container div--> thing. This then opens door to adding other functionality in the future without much work.

Reply:
I always fear the thought of ending up with a monolithic class that contains every possible thing imaginable

Reply:
lol, that would be awesome :)
It's already a nice class. Decision is yours.